tsdb/record/record.go

```diff
@@ -244,6 +332,29 @@ func (e *Encoder) Series(series []RefSeries, b []byte) []byte {
 	return buf.Get()
 }
 
+// Metadata appends the encoded metadata to b and returns the resulting slice.
+func (e *Encoder) Metadata(metadata []RefMetadata, b []byte) []byte {
+	buf := encoding.Encbuf{B: b}
+	buf.PutByte(byte(Metadata))
+
+	for _, m := range metadata {
+		buf.PutUvarint64(uint64(m.Ref))
+
+		// Use a temporary buffer to get the size of the fields
+		// that we're going to encode before writing the fields themselves.
+		tmp := encoding.Encbuf{B: []byte{}}
```

Comment:
How about we re-use the byte slice for the tmp buffer?

Reply:
In this case I was referring to this reuse:

```go
// Metadata appends the encoded metadata to b and returns the resulting slice.
func (e *Encoder) Metadata(metadata []RefMetadata, b []byte) []byte {
	buf := encoding.Encbuf{B: b}
	buf.PutByte(byte(Metadata))

	tmpBuf := make([]byte, 0, 50)

	for _, m := range metadata {
		buf.PutUvarint64(uint64(m.Ref))

		// Use a temporary buffer to get the size of the fields
		// that we're going to encode before writing the fields themselves.
		tmp := encoding.Encbuf{B: tmpBuf[:0]}
		tmp.PutByte(m.Type)
		tmp.PutUvarintStr(string(m.Unit))
		tmp.PutUvarintStr(string(m.Help))

		// Write the size of the temporary buffer and copy back the encoded fields into place.
		buf.PutUvarint(tmp.Len())
		buf.B = append(buf.B, tmp.B...)
		tmpBuf = tmp.B[:0]
	}

	return buf.Get()
}
```

Basically not a new slice for every loop iteration. This is not that unreadable and in practice we usually see benefits of re-using buffers like this in hot path.

Reply:
Oh yeah, that makes sense nice catch! ^^